File: src/qr.ts
import QRCode from "qrcode";

export interface SlintImageData {
  width: number;
  height: number;
  data: Uint8ClampedArray;
}

// Renders 1 pixel per QR module (plus quiet zone); the Slint side scales it
// up with image-rendering: pixelated for crisp edges.
export function qrToImageData(text: string): SlintImageData {
  const qr = QRCode.create(text, { errorCorrectionLevel: "M" });
  const size = qr.modules.size;
  const margin = 4;
  const dim = size + margin * 2;
  const data = new Uint8ClampedArray(dim * dim * 4).fill(255);
  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      if (qr.modules.get(x, y)) {
        const i = ((y + margin) * dim + (x + margin)) * 4;
        data[i] = 0;
        data[i + 1] = 0;
        data[i + 2] = 0;
      }
    }
  }
  return { width: dim, height: dim, data };
}

export const EMPTY_IMAGE: SlintImageData = {
  width: 1,
  height: 1,
  data: new Uint8ClampedArray(4),
};

File: src/qr.rs
// Renders 1 pixel per QR module (plus quiet zone); the Slint side scales
// it up with image-rendering: pixelated for crisp edges.
use qrcode::{EcLevel, QrCode};
use slint::{Rgba8Pixel, SharedPixelBuffer};

pub fn qr_image(text: &str) -> slint::Image {
    match QrCode::with_error_correction_level(text, EcLevel::M) {
        Ok(code) => {
            let size = code.width();
            let margin = 4usize;
            let dim = size + margin * 2;
            let mut buf = SharedPixelBuffer::<Rgba8Pixel>::new(dim as u32, dim as u32);
            let pixels = buf.make_mut_slice();
            pixels.fill(Rgba8Pixel { r: 255, g: 255, b: 255, a: 255 });
            for y in 0..size {
                for x in 0..size {
                    if code[(x, y)] == qrcode::Color::Dark {
                        let i = (y + margin) * dim + (x + margin);
                        pixels[i] = Rgba8Pixel { r: 0, g: 0, b: 0, a: 255 };
                    }
                }
            }
            slint::Image::from_rgba8(buf)
        }
        Err(_) => empty_image(),
    }
}

pub fn empty_image() -> slint::Image {
    slint::Image::from_rgba8(SharedPixelBuffer::<Rgba8Pixel>::new(1, 1))
}
